go/chat/attachment_s3.go

```diff
@@ -0,0 +1,292 @@
+	if err != nil {
+		a.log.Debug("Multi error: %s", err)
+		return "", err
+	}
+
+	var previousParts map[int]s3.Part
+	if previous != nil {
+		previousParts = make(map[int]s3.Part)
+		list, err := multi.ListParts(ctx)
+		if err != nil {
+			a.log.Debug("ignoring multi.ListParts error: %s", err)
+		} else {
+			for _, p := range list {
+				previousParts[p.N] = p
+			}
+		}
+	}
+
+	// need to use ectx in everything in eg.Go() funcs since eg
+	// will cancel ectx in eg.Wait().
+	eg, ectx := errgroup.WithContext(ctx)
+	blockCh := make(chan job)
+	retCh := make(chan s3.Part)
+	eg.Go(func() error {
+		defer close(blockCh)
+		return a.makeBlockJobs(ectx, r, blockCh)
+	})
+	for i := 0; i < a.pipelineSize; i++ {
+		eg.Go(func() error {
+			for b := range blockCh {
+				if err := a.uploadPart(ectx, task, b, previous, previousParts, multi, retCh); err != nil {
+					return err
+				}
+			}
+			return nil
+		})
+	}
+
+	go func() {
+		eg.Wait()
+		close(retCh)
+	}()
+
+	var complete int64
+	var parts []s3.Part
+	for p := range retCh {
+		parts = append(parts, p)
+		complete += p.Size
+		if task.Progress != nil {
+			task.Progress(int(complete), int(size))
+		}
+	}
+	if err := eg.Wait(); err != nil {
+		return "", err
+	}
+
+	a.log.Debug("s3 putMulti all parts uploaded, completing request")
+
+	if err = multi.Complete(ctx, parts); err != nil {
+		a.log.Debug("multi.Complete error: %s", err)
+		return "", err
+	}
+	a.log.Debug("s3 putMulti success, %d parts", len(parts))
+
+	return task.S3Params.ObjectKey, nil
+}
+
+type job struct {
+	block []byte
+	index int
+}
+
+func (a *AttachmentStore) makeBlockJobs(ctx context.Context, r io.Reader, blockCh chan job) error {
+	var partNumber int
+	for {
+		partNumber++
+		block := make([]byte, blockSize)
+		// must call io.ReadFull to ensure full block read
+		n, err := io.ReadFull(r, block)
+		// io.ErrUnexpectedEOF will be returned for last partial block,
+		// which is ok.
+		if err != nil && err != io.ErrUnexpectedEOF && err != io.EOF {
+			return err
+		}
+		if n < blockSize {
+			block = block[:n]
+		}
+		if n > 0 {
+			select {
+			case blockCh <- job{block: block, index: partNumber}:
+			case <-ctx.Done():
+				return ctx.Err()
+			}
+		}
+		if err == io.EOF || err == io.ErrUnexpectedEOF {
+			break
+		}
+	}
+	return nil
+
+}
+
+func (a *AttachmentStore) uploadPart(ctx context.Context, task *UploadTask, b job, previous *AttachmentInfo, previousParts map[int]s3.Part, multi s3.MultiInt, retCh chan s3.Part) error {
+	a.log.Debug("start: upload part %d", b.index)
+	md5sum := md5.Sum(b.block)
+	md5hex := hex.EncodeToString(md5sum[:])
+	if previous != nil {
+		// check s3 previousParts for this block
+		p, ok := previousParts[b.index]
+		if ok && int(p.Size) == len(b.block) && p.ETag == `"`+md5hex+`"` {
+			a.log.Debug("part %d already uploaded to s3", b.index)
+			// check our own previous record for this part
+			ok, err := StashVerifyPart(task.plaintextHash, task.ConversationID, b.index, md5hex)
```

Comment:
It looks like we're only checking the consistency of a ciphertext block in the case where Amazon claims they have it. I don't think that's ok. I think `StashVerifyPart` needs to be the very first thing we do with every new block, using just what's in our own stash, regardless of what Amazon is telling us.

Right now we're reading ciphertext blocks on line 200 and checking them (though not in all cases) on line 234. Every line of code we put between those two things feels like a risk to me. Could we make the reading and the checking happen in one function?